uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/BarChartPopupMediaItem.cpp

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ *  Copyright 2012-2025 Esri
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "BarChartPopupMediaItem.h"
+
+// Qt headers
+#include <QBarSet>
+#include <QColor>
+
+// Maps SDK headers
+#include <PopupMedia.h>
+#include <PopupMediaItem.h>
+#include <PopupMediaValue.h>
+
+namespace Esri::ArcGISRuntime::Toolkit {
+
+/*!
+  \class Esri::ArcGISRuntime::Toolkit::BarChartPopupMediaItem
+  \inmodule ArcGISRuntimeToolkit
+  \brief Represents the contents of a \c{MediaPopupElement::media} to be used for a BarSeries in QML.
+  */
+
+/*!
+  \brief Constructor
+  \list
+    \li \a popupMedia - The \l {Esri::ArcGISRuntime::PopupMedia} {PopupMedia} used to populate the view.
+    \li \a jsonColorArray - The JSON array containing the color information for each bar set.
+    \li \a parent - The optional parent QObject.
+  \endlist
+ */
+BarChartPopupMediaItem::BarChartPopupMediaItem(PopupMedia* popupMedia, const QJsonArray& jsonColorArray, QObject* parent)
+  : PopupMediaItem{popupMedia, parent},
+    m_jsonArr{jsonColorArray}
+{
+}
+
+/*!
+  \brief Returns the list of \c QBarSet objects. You must take ownership of the returned QBarSets.
+ */
+QList<QBarSet*> BarChartPopupMediaItem::barSets()
+{
+  QList<QBarSet*> barSets;
+  auto mediaValue = popupMediaItem()->value();
+
+  for (int i = 0; i < mediaValue->data().length(); i++)
+  {
+    const auto label = mediaValue->labels().at(i);
+    const auto value = mediaValue->data().at(i).toReal();
+
+    if (value >= m_maxValue)
+      m_maxValue = value;
+
+    if (value <= m_minValue)
+      m_minValue = value;
+
+    const auto barset = new QBarSet(label, this);
```

Comment:
Why does the doc state `You must take ownership of the returned QBarSets.`? They have a parent.

Reply:
I was undecided on if I wanted to give these a parent or not since the Series takes ownership. If you look at `PieChartPopupMediaItem` you can see i didn't give that one a parent.

I was concerned over this getting leaked potentially but giving it a parent covers that.

Reply:
If you are confident that the QML side takes ownership, I wouldn't provide any parent. Having a parent means that there are two owners, and if the BarChartPopupMediaItem goes away first, the QML side will have dangling pointers and could crash when it tries to delete them.

I think I commented elsewhere, but I would consider providing no parent at all, and make the method private so it can solely be executed by QML code via the property. That way it's clear that C++ code can't even invoke it (without some hacks, but the intent is clear).

Having a parent is a good insurance policy in most cases, but it could cause crash on exit types of problems that could be hard to diagnose.

Reply:
> If you are confident that the QML side takes ownership

Yes BarSeries and PieSeries take ownership 100%. 

> I would consider providing no parent at all, and make the method private so it can solely be executed by QML code via the property

Done!

> pointers and could crash when it tries to delete them.

Ahh that's a good point as I was running into the reverse of that situation.

Thank you for the recommendation @JamesMBallard .